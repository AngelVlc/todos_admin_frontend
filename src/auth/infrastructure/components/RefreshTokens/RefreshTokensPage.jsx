import React, { useEffect, useState, useContext, useCallback } from "react";
import { Link } from "react-router-dom";
import { TableWithSelectors } from "../../../../shared/infrastructure/components/TableWithSelectors/TableWithSelectors";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import {
  GetRefreshTokensUseCase,
  DeleteSelectedRefreshTokensUseCase,
} from "../../../application/refreshTokens";

export const RefreshTokensPage = () => {
  const [tokens, setTokens] = useState([]);
  const { useCaseFactory } = useContext(AppContext);
  const [pageInfo, setPageInfo] = useState({
    pageNumber: 1,
    pageSize: 10,
    sortColumn: "id",
    sortOrder: "asc",
  });

  const getRefreshTokens = useCallback(async () => {
    const getRefreshTokensUseCase = useCaseFactory.get(GetRefreshTokensUseCase);
    const refreshTokens = await getRefreshTokensUseCase.execute(pageInfo);

    setTokens(refreshTokens);
  }, [useCaseFactory, pageInfo]);

  useEffect(() => {
    getRefreshTokens();
  }, [getRefreshTokens]);

  const onDeleteSelectedTokens = async () => {
    const deleteSelectedRefreshTokensUseCase = useCaseFactory.get(
      DeleteSelectedRefreshTokensUseCase
    );
    const ok = await deleteSelectedRefreshTokensUseCase.execute(tokens);
    if (ok) {
      getRefreshTokens();
    }
  };

  return (
    <div className="container">
      <h3 className="title">Refresh Tokens</h3>
      <nav className="breadcrumb" aria-label="breadcrumbs">
        <ul>
          <li>
            <a href="/">Home</a>
          </li>
          <li className="is-active">
            <Link aria-current="page" to={`/refreshtokens`}>
              Refresh Tokens
            </Link>
          </li>
        </ul>
      </nav>
      <div className="buttons">
        <button
          className="button is-danger"
          data-testid="deleteSelected"
          onClick={() => onDeleteSelectedTokens()}
        >
          Delete Selected
        </button>
      </div>
      <TableWithSelectors
        columnTitles={["ID", "UserID", "Expiration Date"]}
        columnNames={["id", "userId", "expirationDate"]}
        rows={tokens}
        idColumnName={"id"}
        selectedColumnName={"selected"}
        onSelectedChanged={setTokens}
      />
    </div>
  );
};
