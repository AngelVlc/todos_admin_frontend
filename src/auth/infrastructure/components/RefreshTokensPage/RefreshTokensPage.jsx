import React, { useEffect, useState, useContext, useCallback } from "react";
import { Breadcrumb } from "../../../../shared/infrastructure/components/Breadcrumb";
import { TableWithSelectors } from "../../../../shared/infrastructure/components/TableWithSelectors/TableWithSelectors";
import { AppContext } from "../../../../shared/infrastructure/contexts";
import {
  GetRefreshTokensUseCase,
  DeleteSelectedRefreshTokensUseCase,
} from "../../../application/refreshTokens";

export const RefreshTokensPage = () => {
  const [tokens, setTokens] = useState();
  const { useCaseFactory } = useContext(AppContext);
  const [pageInfo, setPageInfo] = useState({
    pageNumber: 1,
    pageSize: 10,
    sortColumn: "id",
    sortOrder: "desc",
  });

  const getRefreshTokens = useCallback(async () => {
    const getRefreshTokensUseCase = useCaseFactory.get(GetRefreshTokensUseCase);
    const refreshTokens = await getRefreshTokensUseCase.execute(pageInfo);

    setTokens(refreshTokens);
  }, [useCaseFactory, pageInfo]);

  useEffect(() => {
    getRefreshTokens();
  }, [getRefreshTokens]);

  const onDeleteSelectedTokens = async () => {
    const deleteSelectedRefreshTokensUseCase = useCaseFactory.get(
      DeleteSelectedRefreshTokensUseCase
    );
    const ok = await deleteSelectedRefreshTokensUseCase.execute(tokens);
    if (ok) {
      getRefreshTokens();
    }
  };

  return (
    <>
      {tokens && (
        <div className="container">
          <Breadcrumb
            items={[{ url: "/refreshtokens", text: "Refresh Tokens" }]}
          />
          <h3 className="title">Refresh Tokens</h3>
          <div className="buttons">
            <button
              className="button is-danger"
              data-testid="deleteSelected"
              onClick={() => onDeleteSelectedTokens()}
            >
              Delete Selected
            </button>
          </div>
          <TableWithSelectors
            columns={[
              { name: "id", title: "ID" },
              { name: "userId", title: "User ID" },
              { name: "expirationDate", title: "Expiration Date" },
            ]}
            rows={tokens}
            idColumnName={"id"}
            selectedColumnName={"selected"}
            paginationInfo={pageInfo}
            changeSelected={setTokens}
            changePagination={setPageInfo}
          />
        </div>
      )}
    </>
  );
};
