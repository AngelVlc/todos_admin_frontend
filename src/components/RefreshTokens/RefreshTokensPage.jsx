import React, { useEffect, useState } from 'react';
import axios from 'axios';

export const RefreshTokensPage = () => {
  const [tokens, setTokens] = useState([]);

  const getRefreshTokens = async () => {
    const res = await axios.get('refreshtokens')
    const data = res.data.map(token => {
      return {
        ...token,
        selected: false
      }
    });
    setTokens(data);
  }

  useEffect(() => {
    getRefreshTokens()
  }, []);

  const onSelectAll = (newValue) => {
    const data = [...tokens].map((token) => {
      return {
        ...token,
        selected: newValue
      }
    });
    setTokens(data);
  };

  const onChangeItemSelected = (newValue, index) => {
    const data = [...tokens];
    data[index].selected = newValue;
    setTokens(data);
  };

  const onDeleteSelectedTokens = async () => {
    const selectedIDs = tokens.reduce((filtered, token) => {
      if (token.selected) {
         filtered.push(token.id);
      }
      return filtered;
    }, []);

    const res = await axios.delete('refreshtokens', {data: selectedIDs});
    if (res.status === 204) {
      getRefreshTokens();
    }
  };

  return (
    <div className="container">
      <h3 className="title">Refresh Tokens</h3>
      <div className="buttons">
        <button className="button is-danger" data-testid="deleteSelected" onClick={() => onDeleteSelectedTokens()}>Delete Selected</button>
      </div>
      <table className="table">
        <thead>
          <tr>
            <td>ID</td>
            <td>UserID</td>
            <td>Expiration Date</td>
            <td>
              <center>
                  <input type="checkbox" data-testid="toggleSelectAll" defaultChecked={false} onChange={(e) => onSelectAll(e.target.checked)}></input>
                </center>
            </td>
          </tr>
        </thead>
        <tbody>
          {tokens.length > 0 && tokens.map((token, index) => (
            <tr key={token.id}>
              <td>{token.id}</td>
              <td>{token.userId}</td>
              <td>{token.expirationDate}</td>
              <td>
                <center>
                  <input type="checkbox" data-testid={`checkBoxItem${token.id}`} checked={token.selected} onChange={(e) => onChangeItemSelected(e.target.checked, index)}></input>
                </center>
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
}
