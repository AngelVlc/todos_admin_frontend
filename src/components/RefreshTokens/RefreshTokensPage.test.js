import { render, cleanup, fireEvent, wait } from '@testing-library/react'
import { RefreshTokensPage } from './RefreshTokensPage';
import { AppContext } from '../../contexts/AppContext'
import axios from 'axios';
import { act } from 'react-dom/test-utils';

afterEach(cleanup)

jest.mock('axios');

const renderWithContext = (component) => {
  axios.get.mockResolvedValue(
      {
          data: [
              { id: 1, userId: 1, expirationDate: '2021-01-29T16:46:58Z' },
              { id: 2, userId: 2, expirationDate: '2021-03-29T16:46:58Z' },
              { id: 3, userId: 1, expirationDate: '2021-04-29T16:46:58Z' }
          ]
      }
  );
  const context = { auth: { info: {} } };
  return {
      ...render(
          <AppContext.Provider value={context}>
              {component}
          </AppContext.Provider>)
  }
}

it('should match the snapshot', async () => {
  let fragment;
  await act(async () => {
      const { asFragment } = renderWithContext(<RefreshTokensPage />);
      fragment = asFragment;
  });
  expect(fragment(<RefreshTokensPage />)).toMatchSnapshot();
});

it('should select all the refresh tokens', async() => {
  let container;
  await act(async () => {
      container = renderWithContext(<RefreshTokensPage />);
  });

  await wait(() => {
      fireEvent.click(container.getByTestId('toggleSelectAll'));
  })

  expect(container.getByTestId('checkBoxItem1')).toBeChecked();
  expect(container.getByTestId('checkBoxItem2')).toBeChecked();
  expect(container.getByTestId('checkBoxItem3')).toBeChecked();
});

it('should unselect all the refresh tokens', async() => {
  let container;
  await act(async () => {
      container = renderWithContext(<RefreshTokensPage />);
  });

  await wait(() => {
      fireEvent.click(container.getByTestId('toggleSelectAll'));
      fireEvent.click(container.getByTestId('checkBoxItem1'));
      fireEvent.click(container.getByTestId('checkBoxItem2'));
      fireEvent.click(container.getByTestId('checkBoxItem3'));
  })

  expect(container.getByTestId('checkBoxItem1')).not.toBeChecked();
  expect(container.getByTestId('checkBoxItem2')).not.toBeChecked();
  expect(container.getByTestId('checkBoxItem3')).not.toBeChecked();
});

it('should unselect all the refresh tokens', async() => {
  let container;
  await act(async () => {
      container = renderWithContext(<RefreshTokensPage />);
  });

  await wait(() => {
      fireEvent.click(container.getByTestId('checkBoxItem1'));
      fireEvent.click(container.getByTestId('checkBoxItem3'));
  })

  axios.delete.mockResolvedValue({ status: 204 });

  axios.get.mockResolvedValue(
    {
        data: [
            { id: 2, userId: 2, expirationDate: '2021-03-29T16:46:58Z' },
        ]
    }
  );

  const cbItem1 = container.getByTestId('checkBoxItem1');
  const cbItem3 = container.getByTestId('checkBoxItem3');

  await wait(() => {
    fireEvent.click(container.getByTestId('deleteSelected'));
  })

  expect(axios.delete.mock.calls.length).toBe(1);
  expect(axios.delete.mock.calls[0][0]).toBe('refreshtokens');
  expect(axios.delete.mock.calls[0][1]).toStrictEqual({ data: [1,3] });
  expect(cbItem1).not.toBeInTheDocument();
  expect(container.getByTestId('checkBoxItem2')).toBeInTheDocument();
  expect(cbItem3).not.toBeInTheDocument();
});