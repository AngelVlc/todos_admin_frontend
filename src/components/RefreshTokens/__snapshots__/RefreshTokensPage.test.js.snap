// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h3
      class="title"
    >
      Refresh Tokens
    </h3>
    <div
      class="buttons"
    >
      <button
        class="button is-danger"
        data-testid="deleteSelected"
      >
        Delete Selected
      </button>
    </div>
    <table
      class="table"
    >
      <thead>
        <tr>
          <td>
            ID
          </td>
          <td>
            UserID
          </td>
          <td>
            Expiration Date
          </td>
          <td>
            <center>
              <input
                data-testid="toggleSelectAll"
                type="checkbox"
              />
            </center>
          </td>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            1
          </td>
          <td>
            1
          </td>
          <td>
            2021-01-29T16:46:58Z
          </td>
          <td>
            <center>
              <input
                data-testid="checkBoxItem1"
                type="checkbox"
              />
            </center>
          </td>
        </tr>
        <tr>
          <td>
            2
          </td>
          <td>
            2
          </td>
          <td>
            2021-03-29T16:46:58Z
          </td>
          <td>
            <center>
              <input
                data-testid="checkBoxItem2"
                type="checkbox"
              />
            </center>
          </td>
        </tr>
        <tr>
          <td>
            3
          </td>
          <td>
            1
          </td>
          <td>
            2021-04-29T16:46:58Z
          </td>
          <td>
            <center>
              <input
                data-testid="checkBoxItem3"
                type="checkbox"
              />
            </center>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
